### ISSUE TITLE
can't merge some dl1 data

### ISSUE THREAD

**Issue by SeiyaNozaki** [cta-observatory/cta-lstchain#290]
I copied the DL1 data in my directory vir symbolic link run by run, then I launched the script to merge all of those files like below.

~~(cta-dev) [seiya.nozaki@cp01 scripts]$ python lstchain_merge_hdf5_files.py \
      --source-dir /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01880 \
      --outfile /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1/dl1_Run01880.fits.h5 \
      --smart False  --no-image True~~

(cta-dev) [seiya.nozaki@cp01 scripts]$ python lstchain_merge_hdf5_files.py \
      --source-dir /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01881 \
      --outfile /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1/dl1_Run01881.fits.h5 \
      --smart False  --no-image True

Then I got the message like below;

/fefs/aswg/workspace/seiya.nozaki/Install/anaconda3/envs/cta-dev/lib/python3.6/site-packages/sklearn/externals/joblib/__init__.py:15: DeprecationWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=DeprecationWarning)
/fefs/aswg/workspace/seiya.nozaki/Install/anaconda3/envs/cta-dev/lib/python3.6/site-packages/tables/path.py:157: NaturalNameWarning: object name is not a valid Python identifier: 'layout.__table_column_meta__'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
  check_attribute_name(name)
**Can't append node dl1/event/telescope/parameters/LST_LSTCam from file /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01881/dl1_LST-1.1.Run01881.0086.fits.h5
Can't append node dl1/event/telescope/parameters/LST_LSTCam from file /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01881/dl1_LST-1.1.Run01881.0119.fits.h5**
.....

So I cannot merge some DL1 files, but not all the files...
And I could merge all of the files for some runs without problems.
Do you have any ideas...?

**Comment by morcuended** [comment]
Thanks for reporting @SeiyaNozaki

Tried to merge run1880 & run1881. I'm also getting those warnings:

```
(cta) [daniel.morcuende@cp02 20200127_link]$  srun lstchain_merge_hdf5_files -d ~/data/real/DL1/20200127_link -o dl1_Run01880.fits.h5 --smart False --no-image True &

/fefs/aswg/software/virtual_env/anaconda3/envs/cta/lib/python3.7/site-packages/corsikaio/subblocks/dtypes.py:20: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  return np.dtype(dict(**dt))
/fefs/aswg/software/virtual_env/anaconda3/envs/cta/lib/python3.7/site-packages/sklearn/externals/joblib/__init__.py:15: FutureWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=FutureWarning)
/fefs/aswg/software/virtual_env/anaconda3/envs/cta/lib/python3.7/site-packages/tables/path.py:155: NaturalNameWarning: object name is not a valid Python identifier: 'layout.__table_column_meta__'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
  check_attribute_name(name)
/fefs/aswg/software/virtual_env/anaconda3/envs/cta/lib/python3.7/site-packages/tables/path.py:155: NaturalNameWarning: object name is not a valid Python identifier: 'LSTCam.__table_column_meta__'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
  check_attribute_name(name)
/fefs/aswg/software/virtual_env/anaconda3/envs/cta/lib/python3.7/site-packages/tables/path.py:155: NaturalNameWarning: object name is not a valid Python identifier: 'optics.__table_column_meta__'; it does not match the pattern ``^[a-zA-Z_][a-zA-Z0-9_]*$``; you will not be able to use natural naming to access this object; using ``getattr()`` will still work, though
  check_attribute_name(name)
```
In the case of run 1881 I also got those append errors:

```
Can't append node dl1/event/telescope/parameters/LST_LSTCam from file /home/daniel.morcuende/data/real/DL1/20200127_link/dl1_LST-1.1.Run01881.0119.fits.h5
Can't append node dl1/event/telescope/parameters/LST_LSTCam from file /home/daniel.morcuende/data/real/DL1/20200127_link/dl1_LST-1.1.Run01881.0099.fits.h5
Can't append node dl1/event/telescope/parameters/LST_LSTCam from file /home/daniel.morcuende/data/real/DL1/20200127_link/dl1_LST-1.1.Run01881.0034.fits.h5
```

Checking the file I realized that UCTS info is missing for some of the subruns.
![image](https://user-images.githubusercontent.com/32680689/74944267-11910b80-53f6-11ea-81a2-c0d1e98ca192.png)

So the problematics columns are `alt_tel` and `az_tel` which in the merged file are missing. Probably related to #285.

I guess one workaround could be to produce DL2 files for each subrun, without merging at DL1 level and then in principle even though pointing coordinates are missing it should not be a problem. What do you think @vuillaut?

**Comment by morcuended** [comment]
> I guess one workaround could be to produce DL2 files for each subrun, without merging at DL1 level and then in principle even though pointing coordinates are missing it should not be a problem. What do you think @vuillaut?

Of course after the new release. Currently, we cannot produce DL2 files with standard scripts due to the bug in pointing units

**Comment by vuillaut** [comment]
> I guess one workaround could be to produce DL2 files for each subrun, without merging at DL1 level and then in principle even though pointing coordinates are missing it should not be a problem. What do you think @vuillaut?

Have you tried?
I am not sure what will happen if you try to merge two datasets with different columns (e.g. one missing `alt_tel`) but I would be (positively) surprised if it works.

**Comment by morcuended** [comment]
No, I haven't yet, but you're right, two files with different columns won't work I guess. However, I was thinking that at least if you have altaz coordinates for some of the events, the rest would be filled with 0:

    if 'mc_alt_tel' in dl2.columns:
        alt_tel = dl2['mc_alt_tel'].values
        az_tel = dl2['mc_az_tel'].values
    elif 'alt_tel' in dl2.columns:
        alt_tel = dl2['alt_tel'].values
        az_tel = dl2['az_tel'].values
    else:
        alt_tel = np.zeros(len(dl2))
        az_tel = np.zeros(len(dl2))

But of course, you need at least have `alt_tel` and `az_tel` columns in the file already.

**Comment by vuillaut** [comment]
> No, I haven't yet, but you're right, two files with different columns won't work I guess. However, I was thinking that at least if you have altaz coordinates for some of the events, the rest would be filled with 0:
> 
> ```
> if 'mc_alt_tel' in dl2.columns:
>     alt_tel = dl2['mc_alt_tel'].values
>     az_tel = dl2['mc_az_tel'].values
> elif 'alt_tel' in dl2.columns:
>     alt_tel = dl2['alt_tel'].values
>     az_tel = dl2['az_tel'].values
> else:
>     alt_tel = np.zeros(len(dl2))
>     az_tel = np.zeros(len(dl2))
> ```
> 
> But of course, you need at least have `alt_tel` and `az_tel` columns in the file already.

Ha, in current master there won't be any issue regarding these two columns since they are forced to be created (the code you copied).

**Comment by rlopezcoto** [comment]
Time for a new bugfix release? :)
@morcuended had already commented on that.

**Comment by vuillaut** [comment]
> In the case of run 1881 I also got those append errors:
> 
> ```
> Can't append node dl1/event/telescope/parameters/LST_LSTCam from file /home/daniel.morcuende/data/real/DL1/20200127_link/dl1_LST-1.1.Run01881.0119.fits.h5
> Can't append node dl1/event/telescope/parameters/LST_LSTCam from file /home/daniel.morcuende/data/real/DL1/20200127_link/dl1_LST-1.1.Run01881.0099.fits.h5
> Can't append node dl1/event/telescope/parameters/LST_LSTCam from file /home/daniel.morcuende/data/real/DL1/20200127_link/dl1_LST-1.1.Run01881.0034.fits.h5
> ```
> 
> 
> So the problematics columns are `alt_tel` and `az_tel` which in the merged file are missing. Probably related to #285.

Are you sure @morcuended ?
Checking `/home/daniel.morcuende/data/real/DL1/20200127_link/dl1_LST-1.1.Run01881.0119.fits.h5`, the columns `alt_tel` and `az_tel` do exist.

```
import pandas as pd
from lstchain.io.io import dl1_params_lstcam_key
df = pd.read_hdf('/home/daniel.morcuende/data/real/DL1/20200127_link/dl1_LST-1.1.Run01881.0119.fits.h5', key=dl1_params_lstcam_key)
In [7]: df.columns
Out[7]:
Index(['alt_tel', 'az_tel', 'dragon_time', 'event_id', 'intensity',
       'intercept', 'kurtosis', 'leakage', 'length', 'log_intensity',
       'mc_core_distance', 'n_islands', 'obs_id', 'phi', 'psi', 'r',
       'skewness', 'tel_id', 'tel_pos_x', 'tel_pos_y', 'tel_pos_z', 'tib_time',
       'time_gradient', 'ucts_time', 'width', 'wl', 'x', 'y'],
      dtype='object')

**Comment by morcuended** [comment]
Well, in some of them I guess they exist, but in the merged file no:

```
filename = "/home/daniel.morcuende/data/real/DL1/20200127_link/dl1_Run01881.fits.h5"
df = pd.read_hdf(filename, key=dl1_params_lstcam_key)
df.columns

Index(['dragon_time', 'event_id', 'intensity', 'intercept', 'kurtosis',
       'leakage', 'length', 'log_intensity', 'mc_core_distance', 'n_islands',
       'obs_id', 'phi', 'psi', 'r', 'skewness', 'tel_id', 'tel_pos_x',
       'tel_pos_y', 'tel_pos_z', 'tib_time', 'time_gradient', 'ucts_time',
       'width', 'wl', 'x', 'y'],
      dtype='object')
```

**Comment by vuillaut** [comment]
> So I cannot merge some DL1 files, but not all the files...
> And I could merge all of the files for some runs without problems.
> Do you have any ideas...?

Hi @SeiyaNozaki 

At least the core function works

```
from lstchain.io import auto_merge_h5files
list =['/fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01881/dl1_LST-1.1.Run01881.0086.fits.h5', '/fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01881/dl1_LST-1.1.Run01881.0119.fits.h5', '/fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01881/dl1_LST-1.1.Run01881.0000.fits.h5']

auto_merge_h5files(list)
```

Still investigating.

**Comment by vuillaut** [comment]
> (cta-dev) [seiya.nozaki@cp01 scripts]$ python lstchain_merge_hdf5_files.py
> --source-dir /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01880
> --outfile /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1/dl1_Run01880.fits.h5
> --smart False --no-image True

```
lstchain_merge_hdf5_files -d /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01880 --outfile merged.h5 --smart False --no-image True
```

works fine.
I get a `merged.h5` file of 2.0GB 🤔

**Comment by morcuended** [comment]
> > (cta-dev) [seiya.nozaki@cp01 scripts]$ python lstchain_merge_hdf5_files.py
> > --source-dir /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01880
> > --outfile /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1/dl1_Run01880.fits.h5
> > --smart False --no-image True
> 
> ```
> lstchain_merge_hdf5_files -d /fefs/aswg/workspace/seiya.nozaki/data/Crab/200127/DL1_raw/Run01880 --outfile merged.h5 --smart False --no-image True
> ```
> 
> works fine.
> I get a `merged.h5` file of 2.0GB

Yes, run 1880 does not have any missing `ucts_time` and therefore all subruns have `alt_tel` column defined.

**Comment by morcuended** [comment]
Maybe we should get rid of those events with missing `ucts_time` right after producing DL1 files.

**Comment by vuillaut** [comment]
> Maybe we should get rid of those events with missing `ucts_time` right after producing DL1 files.
> 
> Edit: as you mention already

But I run the command @SeiyaNozaki was trying to do, so at least his bug is something else?

Concerning the `ucts_time` error events, I agree they should be treated in a way or another.

**Comment by SeiyaNozaki** [comment]
Oops, sorry...
I copy&paste the commands and error outputs for different runs...
Actually I got this error for run1881, not run1880.
Sorry for the confusion...!

**Comment by SeiyaNozaki** [comment]
Just let me clarify the solution following the discussion in #295 , so merging after DL2 is recommended, right?

**Comment by rlopezcoto** [comment]
Solved in #295. 
@SeiyaNozaki Merging at DL1 level should also be fine after all the columns for `az_tel` and `alt_tel` are written regardless of the `ucts_time` information being present or not.
